### PR TITLE
port atan2 from TH to ATen

### PR DESCRIPTION
**Pull request by ifedan**
#22799

BC: enforce stronger shape requirements on the output tensor (out= keyword argument) and do not allow output tensor to be resized if it is also used as one of the inputs.

### CODE REVIEW

File: aten/src/ATen/native/BinaryOps.cpp [hunk 1]
```diff
@@ -151,6 +152,28 @@ Tensor rsub(const Tensor& self, const Tensor& other, Scalar alpha) {
   return native::sub(other, self, alpha);
 }
 
+
+Tensor& atan2_out(Tensor& result, const Tensor& self, const Tensor& other) {
+  auto device_res = result.type().device_type();
+  auto device1 = self.type().device_type();
+  TORCH_CHECK(device_res == device1, "out and input1 must have the same device. out: ", device_res, " input1: ", device1);
```
Comment (gchanan) [inline aten/src/ATen/native/BinaryOps.cpp]: does it make any sense to move the checking to TensorIterator itself?

Comment (VitalyFedyunin) [reply aten/src/ATen/native/BinaryOps.cpp]: It is already inside of thr TensorIterator

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -3322,8 +3322,8 @@
 - func: atan2_(Tensor(a!) self, Tensor other) -> Tensor(a!)
   variants: method
   dispatch:
-    CPU: legacy::cpu::_th_atan2_
-    CUDA: legacy::cuda::_th_atan2_
+    CPU: atan2_
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/native_functions.yaml]: No need to define dispatch if all function names matches.

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Avoid duplicate device checks.

**Comment by VitalyFedyunin** [comment]
Please attach performance benchmarks for cpu and cuda.

**Comment by ifedan** [comment]
> Please attach performance benchmarks for cpu and cuda.
**test**
```
import timeit
import torch
import itertools
import statistics

def test_perf(sizes, device, repeat, times):
    def _tensor(name, sizes, device):
        return '''{0} = torch.rand({1}, device="{2}");'''.format(name, sizes, device)

    setup_code = 'import torch;' + _tensor('x', sizes, device) + _tensor('y', sizes, device)
    test_code = '''torch.atan2(y, x);'''
    if device == "cuda":
        test_code = test_code + 'torch.cuda.synchronize()'
    result = timeit.repeat(setup = setup_code,stmt = test_code,repeat = repeat,number = times)
    mean = statistics.mean(result)
    std = statistics.stdev(result)
    print('''sizes = {0} mean = {1} std = {2}'''.format(sizes, std, mean))

def test_perf_for_device(device, small, mid, large):
    print(device)
    for s in itertools.product((small, mid, large), (small, mid, large)):
        test_perf(str(s), device, 3, 300)

test_perf_for_device("cpu", 5, 100, 1000)
test_perf_for_device("cuda", 5, 100, 10000)
```
**pytorch:master**
```
cpu
sizes = (5, 5) mean = 0.0001450333580075422 std = 0.0017364399585252006
sizes = (5, 100) mean = 0.0029784211023988665 std = 0.013106301698523263
sizes = (5, 1000) mean = 0.03438556826284164 std = 0.09237984466987352
sizes = (100, 5) mean = 2.478881328151161e-05 std = 0.008162023305582503
sizes = (100, 100) mean = 0.04055429240048574 std = 0.18722845831265053
sizes = (100, 1000) mean = 0.1260943765593662 std = 1.6641836036384727
sizes = (1000, 5) mean = 0.015972869647852585 std = 0.08150684131154169
sizes = (1000, 100) mean = 0.232198996201533 std = 1.789468562696129
sizes = (1000, 1000) mean = 0.3790670978868011 std = 15.708918012639819
cuda
sizes = (5, 5) mean = 0.007155040446700871 std = 0.024204802660581965
sizes = (5, 100) mean = 0.0010435590256651479 std = 0.012264323653653264
sizes = (5, 10000) mean = 0.00030993165320112524 std = 0.011815296291994551
sizes = (100, 5) mean = 0.0011544815538170204 std = 0.012403536005876958
sizes = (100, 100) mean = 0.0024333436330361375 std = 0.012738600994149843
sizes = (100, 10000) mean = 0.00022303123519843825 std = 0.025823221658356488
sizes = (10000, 5) mean = 0.0003053356125067329 std = 0.011981462361291051
sizes = (10000, 100) mean = 0.00038227427138423707 std = 0.026194119709543884
sizes = (10000, 10000) mean = 0.14192976771988677 std = 1.8195964709933226
```
**ifedan:atan2**
```
cpu
sizes = (5, 5) mean = 0.0011233232630492738 std = 0.0016390623447174828
sizes = (5, 100) mean = 5.818948128526334e-05 std = 0.0017121707011635106
sizes = (5, 1000) mean = 0.001347213862655638 std = 0.013473411711553732
sizes = (100, 5) mean = 9.46000878180635e-05 std = 0.0032052043825387955
sizes = (100, 100) mean = 5.3258892426376576e-05 std = 0.01503851703212907
sizes = (100, 1000) mean = 0.0011398015164726196 std = 0.013505248275275031
sizes = (1000, 5) mean = 5.0381620252363384e-05 std = 0.008026461970681945
sizes = (1000, 100) mean = 1.2381679033686386e-05 std = 0.012970551343945166
sizes = (1000, 1000) mean = 0.000256763887236208 std = 0.10153608365605275
cuda
sizes = (5, 5) mean = 0.000163014257207611 std = 0.007157436067548891
sizes = (5, 100) mean = 2.133623326224886e-05 std = 0.007204662348764638
sizes = (5, 10000) mean = 3.7335672042653244e-05 std = 0.007460295998801787
sizes = (100, 5) mean = 0.00010144182686921984 std = 0.00725937068151931
sizes = (100, 100) mean = 0.0001658543542412084 std = 0.007338125646735231
sizes = (100, 10000) mean = 5.003247881338404e-05 std = 0.022574760058584314
sizes = (10000, 5) mean = 7.441024064507601e-05 std = 0.007160835006895165
sizes = (10000, 100) mean = 7.081770102160384e-05 std = 0.02249801465465377
sizes = (10000, 10000) mean = 0.0004216414246337847 std = 1.5984146439392741
```

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (ifedan) [reply aten/src/ATen/native/native_functions.yaml]: Fixed

File: aten/src/ATen/native/BinaryOps.cpp (hunk 1 above)
Comment (ifedan) [reply aten/src/ATen/native/BinaryOps.cpp]: Fixed

**Comment by ifedan** [comment]
> Avoid duplicate device checks.

Fixed

**Comment by ifedan** [comment]
> Please attach performance benchmarks for cpu and cuda.

Added

**Comment by fmassa** [comment]
@ifedan the variances on your benchmark seem way too high, looks like you inverted the mean and the std in the print?

**Comment by ifedan** [comment]
> @ifedan the variances on your benchmark seem way too high, looks like you inverted the mean and the std in the print?

Good catch! Fixed.

**Comment by VitalyFedyunin** [comment]
If you want to cover as much as possible you should also add broadcasting checks

**Comment by ifedan** [comment]
> If you want to cover as much as possible you should also add broadcasting checks

They are already in place

**Review by VitalyFedyunin — DISMISSED** [review]

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
edgecases test is failing atan2 checks

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
enforce stronger shape requirements on the output tensor (out= keyword argument) and do not allow output tensor to be resized if it is also used as one of the inputs.

**Pull request closed without merging**